### PR TITLE
fix RAF bindings on IE11

### PR DESCRIPTION
**Pull request by amireh**
refs GH-99

Maintaining a reference to "requestAnimationFrame" and friends
(including setTimeout) without binding them to the window object seems
to make IE angry with an exception similar to:

    TypeError: Invalid calling object

Binding the function references to window fixes the issue.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -10,10 +10,13 @@ import uid from './uid';
...
-  window.requestAnimationFrame
-  ? [window.requestAnimationFrame, window.cancelAnimationFrame]
-  : [setTimeout, clearTimeout];
+const [onNextFrame, clearNextFrameAction] =
+  isBrowser && window.requestAnimationFrame
+    ? [
+      window.requestAnimationFrame.bind(window),
+      window.cancelAnimationFrame.bind(window),
+    ]
+    : [window.setTimeout.bind(window), window.clearTimeout.bind(window)];
```
Comment (Andarist) [inline src/index.js]: Library might be used outside of the browser context (SSR, tests), so binding with `window` is not the best idea. Better to just create a proxy functions like:
```js
(fn, ms) => setTimeout(fn, ms)
```

Comment (amireh) [reply src/index.js]: Done

Comment (amireh) [reply src/index.js]: edit: in fact i removed that part of the change altogether since it's not necessary (no browser context = no ie, no need)

**Comment by amireh** [comment]
Any chance we can get this in soon? We'd rather not keep using a fork / github in npm.

Thanks!

**Comment by Andarist** [comment]
@amireh could u provide steps for reproduction? also your configuration would be helpful + webpack etc versions

Ive just setup a simple create-react-app project, tested it on IE11 and I have no issue with this.

**Comment by cainlevy** [comment]
Confirming that this patch solves the issue I'm experiencing with IE11.

**Comment by Andarist** [comment]
Without having exact steps for reproduction this PR has **zero** chances of being merged in. 

I want to be sure that the change is needed and that the issue originates here in the library and not in a tooling like webpack.

Does the problem occurs in dev or prod mode? What version of webpack do u use? What type of source maps? Etc, etc.

**Comment by cainlevy** [comment]
I was able to finally reproduce this with `create-react-app` by ejecting and changing `devtool` to `eval-source-map` in `config/webpack.config.dev.js`:

1. make something tiny with create-react-app and react-textarea-autosize
2. eject
3. change `config/webpack.config.dev.js` to use `devtool: 'eval-source-map'`
4. load app in IE11
5. attempt to type in the textarea

**Comment by Andarist** [comment]
Seems to be the same case as [this](https://github.com/andreypopp/react-textarea-autosize/pull/170#issuecomment-330994808). 

I don't intend to support IE quirk in **development mode only**, especially if it only happens when using "eval" source maps.

**Comment by juhoha** [comment]
@Andarist I'd ask you to reconsider. Amireh's changes conform to better practices anyway. Using a forked version here as well: a slow devtool for a 30mb bundle isn't feasible for one dependency. A problem this small spawning a long-lived fork just seems silly.

**Comment by Andarist** [comment]
I don't think that it really conforms to better practices, those functions are this-less and can be passed around as reference safely - the fact that IE has it bugged doesn't change anything.

However after a consideration I can add a development branch of this code to support this broken 💩.

**Comment by cainlevy** [comment]
A big part of the value proposition for browser libraries is abstracting across browser differences. These are the quirks that shouldn't waste anyone's time, whether in development or in production. It's an imperfect situation that we have collectively been trying to make the best of for over a decade.

If the only fix for this issue is within a development branch, then it creates a discoverability issue and will continue to waste developers' time.

At the very least, please consider updating the compatibility statement in the readme:

> This module supports IE9 and above.

**Comment by Andarist** [comment]
As the issue appears to be only relevant for using `eval` in IE11 (and I thought we've also as community established that using `eval` should be avoided), I think it's a reasonable thing to make the development branch different than production one.

Feel free to raise your concerns further, I'm also just figuring this thing out and as english is not my mother tongue I think my statements might sound a little bit rude and harsh - if so, please do not take them as such, this is not my intention.

**Comment by cainlevy** [comment]
Thanks for keeping the discussion open! Also, thanks for maintaining a useful open source project. I want to remain respectful of your time and your ability to decide what's best for the project.

Do you have any examples in mind of other projects with development branches? How do people discover the branch, configure an app to use the correct branch in the correct environment, and gain confidence that what they're deploying has been tested?

When you first suggested a development branch, I guessed it would behave like an official fork. I may have misunderstood.

**Comment by Andarist** [comment]
Oh, by the development branch I mean something like this:
```js
if (process.env.NODE_ENV === 'development') {
  // ...
} else {
  // ...
}
```
It's a commonly used practice for creating DEV only code. `process.env.NODE_ENV` gets replaced during build process and dead code branches are later removed by minifiers.

**Comment by cainlevy** [comment]
Thanks @Andarist, that sounds like it would solve the problem. I do hope no one is using `eval` in production. 😃

**Comment by juhoha** [comment]
I thought you meant a git branch/fork as well. But yeah, flagging the functionality behind env would definitely solve the problem :+1:

**Comment by Andarist** [comment]
Sorry @amireh that it took so long and that finally your PR didn't get merged. I've just wanted to make changes quickly, so I could release it right now.

**Pull request closed without merging**